From 51c0845cbab7bcd541773e84229fbfaacfb1f518 Mon Sep 17 00:00:00 2001
From: Anil Dsouza <anil.dsouza@einfochips.com>
Date: Mon, 21 Mar 2022 12:09:41 +0530
Subject: [PATCH 1/2] Add support for ar1335 camera sensor interface using
 AP1302 ISP

*) Added the support for AR1335 camera sensor in AP1302 driver
*) New dts file created for AR1335 camera sensor support
*) AR1335 device tree overlay changes added in the dts Makefile
*) Increase the resolution in dcmi driver to support 13mp capture

Change-Id: I64714148728e7403232eb071343f427854d71efc
Signed-off-by: Anil Dsouza <anil.dsouza@einfochips.com>
---
 arch/arm/boot/dts/Makefile                    |  1 +
 ...mp15xx-avenger96-overlay-ap1302-ar1335.dts | 95 +++++++++++++++++++
 drivers/media/i2c/ap1302.c                    |  5 +-
 drivers/media/platform/stm32/stm32-dcmi.c     |  4 +-
 4 files changed, 101 insertions(+), 4 deletions(-)
 create mode 100644 arch/arm/boot/dts/stm32mp15xx-avenger96-overlay-ap1302-ar1335.dts

diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index 91b9d4844..85311506d 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -1076,6 +1076,7 @@ dtb-$(CONFIG_ARCH_STM32) += \
 	stm32mp15xx-avenger96-overlay-spi2-eeprom-x6.dtbo \
 	stm32mp15xx-avenger96-overlay-ap1302-ar0430.dtbo \
 	stm32mp15xx-avenger96-overlay-ap1302-arx3a0.dtbo \
+	stm32mp15xx-avenger96-overlay-ap1302-ar1335.dtbo \
 	stm32mp157a-dk1.dtb \
 	stm32mp157a-iot-box.dtb \
 	stm32mp157a-stinger96.dtb \
diff --git a/arch/arm/boot/dts/stm32mp15xx-avenger96-overlay-ap1302-ar1335.dts b/arch/arm/boot/dts/stm32mp15xx-avenger96-overlay-ap1302-ar1335.dts
new file mode 100644
index 000000000..d13a114f5
--- /dev/null
+++ b/arch/arm/boot/dts/stm32mp15xx-avenger96-overlay-ap1302-ar1335.dts
@@ -0,0 +1,95 @@
+#include <dt-bindings/clock/stm32mp1-clks.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/dts-v1/;
+/plugin/;
+
+&{/} {
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	sram@10050000 {
+               compatible = "mmio-sram";
+               reg = <0x10050000 0x10000>;
+               #address-cells = <1>;
+               #size-cells = <1>;
+               ranges = <0 0x10050000 0x10000>;
+
+               dma_pool: dma_pool@0 {
+                       reg = <0x0 0x10000>;
+                       pool;
+               };
+         };
+};
+
+&dma1 {
+	sram = <&dma_pool>;
+};
+
+&dma2 {
+	sram = <&dma_pool>;
+};
+
+&adc {
+	status = "disabled";
+};
+
+&adc1{
+	status = "disabled";
+};
+
+&adc2{
+	status = "disabled";
+};
+
+&i2c2 {
+	ap1302: camera@3d {
+            compatible = "on,ap1302";
+            reg = <0x3d>;
+            pinctrl-names = "default", "sleep";
+            pinctrl-0 = <&rcc_pins_a>;
+            pinctrl-1 = <&rcc_sleep_pins_a>;
+            clocks = <&rcc CK_MCO1>;
+            clock-names = "xclk";
+            assigned-clocks = <&rcc CK_MCO1>;
+            assigned-clock-rates = <24000000>;
+            reset-gpios = <&gpioc 3 GPIO_ACTIVE_LOW>;
+            standby-gpios = <&gpiod 14 GPIO_ACTIVE_HIGH>;
+            i2csel-gpios = <&gpiof 12 GPIO_ACTIVE_HIGH>;
+            isptrig-gpios = <&gpioa 12 GPIO_ACTIVE_HIGH>;
+            status = "okay";
+
+	    port {
+                    ap1302_0: endpoint {
+                            remote-endpoint = <&stmipi_0>;
+                            data-lanes = <1 2>;
+                    };
+            };
+
+	    sensors {
+                #address-cells = <1>;
+                #size-cells = <0>;
+		onnn,model = "onnn,ar1335";
+                    sensor@0 {
+                        reg = <0>;
+                   };
+            };
+      };
+};
+
+&stmipi{
+	status = "okay";
+
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		port@0 {
+			reg = <0>;
+			stmipi_0: endpoint {
+				data-lanes = <1 2>;
+				remote-endpoint = <&ap1302_0>;
+			};
+		};
+	};
+};
diff --git a/drivers/media/i2c/ap1302.c b/drivers/media/i2c/ap1302.c
index 8e681b2d8..051acf3a6 100644
--- a/drivers/media/i2c/ap1302.c
+++ b/drivers/media/i2c/ap1302.c
@@ -545,8 +545,9 @@ static const struct ap1302_sensor_info ap1302_sensor_info[] = {
 		.model = "onnn,ar1335",
 		.name = "ar1335",
 		.i2c_addr = 0x36,
-		.resolution = { 4208, 3120 },
-		.format = MEDIA_BUS_FMT_SGRBG10_1X10,
+		.resolution = { 1280, 720},
+		.fps = 20,
+		.format = MEDIA_BUS_FMT_UYVY8_2X8,
 		.supplies = (const struct ap1302_sensor_supply[]) {
 			{ "vaa", 0 },
 			{ "vddio", 0 },
diff --git a/drivers/media/platform/stm32/stm32-dcmi.c b/drivers/media/platform/stm32/stm32-dcmi.c
index fd1c41cba..12f9a9f41 100644
--- a/drivers/media/platform/stm32/stm32-dcmi.c
+++ b/drivers/media/platform/stm32/stm32-dcmi.c
@@ -91,9 +91,9 @@ enum state {
 };
 
 #define MIN_WIDTH	16U
-#define MAX_WIDTH	2592U
+#define MAX_WIDTH	4208U
 #define MIN_HEIGHT	16U
-#define MAX_HEIGHT	2592U
+#define MAX_HEIGHT	3120U
 
 #define TIMEOUT_MS	1000
 
-- 
2.17.1

